### PR TITLE
enh: Bidirectional mask with multiple images for Gemma3

### PR DESCRIPTION
**Pull request by brb-nv**
## Description

This MR:
- Updates masking utilities needed for context phase requests with image tokens in Gemma3 VLMs. Specifically, when there are multiple images in the same request, tokens of the same image attend to each other.
- Previously, all image tokens attend to each other (regardless of them coming from different images).
- Core change is an algorithm to identify contiguous blobs of ones separated by zeros.

This MR has a nice visualization of the attention mask for global attention and sliding window attention:
https://github.com/huggingface/transformers/pull/38295

Background about custom mask:
- The function `get_flashinfer_attention_mask` will only be called for a batch when there's at least one context request in the batch with image tokens.
- In context phase, each sample's `input_ids` may have a mix of `image` ([image_token_idx](https://huggingface.co/google/gemma-3-4b-it/blob/main/config.json#L11)) and `text` tokens where tokens corresponding to an image appear as a contiguous blob.
Example: torch.IntTensor([2, 3, 4, 5, img_idx, img_idx, img_idx, ..., img_idx, 100])
- While the text tokens attend to other tokens in a causal fashion, image tokens attend to others in a causal fashion and well as attend to other image tokens in a bidirectional manner. Hence, the need for custom masking.

## Test Coverage
```
$ pytest tests/unittest/_torch/modeling/test_modeling_gemma3.py::TestGemma3::test_gemma3_flashinfer_mask -s -v
$ pytest tests/unittest/_torch/modeling/test_modeling_gemma3.py::TestGemma3::test_gemma3_global_context_mask_multi_image -s -v
$ pytest tests/unittest/_torch/modeling/test_modeling_gemma3.py::TestGemma3::test_gemma3_global_context_mask_single_image -s -v
$ pytest tests/unittest/_torch/modeling/test_modeling_gemma3.py::TestGemma3::test_gemma3_local_context_mask_multi_image -s -v
$ pytest tests/unittest/_torch/modeling/test_modeling_gemma3.py::TestGemma3::test_gemma3_local_context_mask_single_image -s -v
```

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

File: tests/unittest/_torch/modeling/test_modeling_gemma3.py [hunk 1]
```diff
@@ -458,53 +459,183 @@ def test_gemma3_global_context_mask(self) -> None:
...
+                # text blob.
+                [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+            ],
+            device=device).bool()
+        torch.testing.assert_close(attention_mask, expected_attention_mask)
+
+        # Even for a sliding window larger than sequence length, the attention mask is the same as the global mask.
+        attention_mask = self.multi_image_mask_test_helper(
+            effective_sliding_window=25)
+        torch.testing.assert_close(attention_mask, expected_attention_mask)
```
Comment (2ez4bz) [inline tests/unittest/_torch/modeling/test_modeling_gemma3.py:594]: Similar comment here re parameterization.

**Review by 2ez4bz — APPROVED** [review]

File: tests/unittest/_torch/modeling/test_modeling_gemma3.py (hunk 1 above)
Comment (brb-nv) [reply tests/unittest/_torch/modeling/test_modeling_gemma3.py]: Done. https://github.com/NVIDIA/TensorRT-LLM/pull/5976/commits/c519db69d709fd5a28924b5bf810903f0eef3654

File: tests/unittest/_torch/modeling/test_modeling_gemma3.py [hunk 2]
```diff
@@ -458,53 +459,183 @@ def test_gemma3_global_context_mask(self) -> None:
...
+             [1, 1, 1, 1, 1, 1, 0, 0], [1, 1, 1, 1, 1, 1, 0, 0],
+             [1, 1, 1, 1, 1, 1, 0, 0], [1, 1, 1, 1, 1, 1, 0, 0],
+             [1, 1, 1, 1, 1, 1, 1, 0], [1, 1, 1, 1, 1, 1, 1, 1]],
+            device=device).bool()
         torch.testing.assert_close(attention_mask, expected_attention_mask)
 
         # Even for a sliding window larger than sequence length, the attention mask is the same as the global mask.
-        attention_mask = gemma3.get_context_mask(
-            image_token_mask=image_token_mask, effective_sliding_window=10)
+        attention_mask = self.single_image_mask_test_helper(
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tests/unittest/_torch/modeling/test_modeling_gemma3.py]: Done. https://github.com/NVIDIA/TensorRT-LLM/pull/5976/commits/c519db69d709fd5a28924b5bf810903f0eef3654

File: tensorrt_llm/_torch/models/modeling_gemma3.py [hunk 3]
```diff
@@ -318,6 +318,41 @@ def __init__(
...
+
+        # Mask out zeros (positions where token_type_ids == 0).
+        token_type_ids = blob_ids * token_type_ids
+
+        # Create a mask where each blob is a unique id.
+        token_type_mask = token_type_ids.unsqueeze(
+            0) == token_type_ids.unsqueeze(1)
+
+        # If text token, do not change anything.
+        token_type_mask[token_type_ids == 0] = False
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/_torch/models/modeling_gemma3.py]: Thank you for sharing benchmark results. I'm surprised by the in-place masked assignment being that expensive. Intuitively, I'd expect to be cheapest. But, clearly, I'm wrong. https://github.com/NVIDIA/TensorRT-LLM/pull/5976/commits/c519db69d709fd5a28924b5bf810903f0eef3654

File: tensorrt_llm/_torch/models/modeling_gemma3.py [hunk 4]
```diff
@@ -318,6 +318,41 @@ def __init__(
...
+
+        # Identify where blobs start (0->1 transitions)
+        starts = (padded[1:] > padded[:-1]).int()
+
+        # Cumulative sum of starts gives a unique id for each blob. Note that
+        # this assigns a unique id to the zeros separating the blobs.
+        blob_ids = torch.cumsum(starts, dim=0)
+
+        # Mask out zeros (positions where token_type_ids == 0).
+        token_type_ids = blob_ids * token_type_ids
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/_torch/models/modeling_gemma3.py]: Done. https://github.com/NVIDIA/TensorRT-LLM/pull/5976/commits/c519db69d709fd5a28924b5bf810903f0eef3654

File: tensorrt_llm/_torch/models/modeling_gemma3.py [hunk 5]
```diff
@@ -318,6 +318,41 @@ def __init__(
                          hidden_size=model_config.pretrained_config.hidden_size,
                          vocab_size=model_config.pretrained_config.vocab_size)
 
+    def _get_token_type_mask(self, image_token_mask: torch.BoolTensor):
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/_torch/models/modeling_gemma3.py:326]: I added this to all `forward()` and `__call__()`. This function is called by a `forward()`. So, that should cover it. Please let me know if you follow a different thumb rule. https://github.com/NVIDIA/TensorRT-LLM/pull/5976/commits/c519db69d709fd5a28924b5bf810903f0eef3654

File: tensorrt_llm/_torch/models/modeling_gemma3.py [hunk 6]
```diff
@@ -318,6 +318,41 @@ def __init__(
...
+        device = image_token_mask.device
+        sequence_length = len(image_token_mask)
+
+        # Create a list of token type ids. 0 for text tokens, 1 for all image tokens (regardless of which image they belong to).
+        token_type_ids = torch.zeros(sequence_length,
+                                     dtype=torch.int32,
+                                     device=device).contiguous()
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/_torch/models/modeling_gemma3.py]: Removed. https://github.com/NVIDIA/TensorRT-LLM/pull/5976/commits/c519db69d709fd5a28924b5bf810903f0eef3654

**Review by byshiue — APPROVED** [review]
